File: ios/SYCCourses/Sources/SYCCourses/Licensing/EntitlementVerifier.swift
import CryptoKit
import Foundation

struct EntitlementVerifier: Sendable {
    let publicKeys: [String: Data]

    func verify(
        _ envelope: EntitlementEnvelope,
        installationId: String,
        expectedClubId: String?,
        expectedPackId: String,
        now: Date,
        allowExpiredReferenceAccess: Bool = true
    ) throws -> VerifiedEntitlement {
        guard let payloadBytes = Data(base64URLEncoded: envelope.payload),
              let signature = Data(base64URLEncoded: envelope.signature)
        else { throw EntitlementVerificationError.malformedEnvelope }

        let decoder = JSONDecoder.licensing
        guard let payload = try? decoder.decode(EntitlementPayload.self, from: payloadBytes)
        else { throw EntitlementVerificationError.malformedEnvelope }
        guard payload.schemaVersion == 1 else { throw EntitlementVerificationError.unknownSchemaVersion }
        guard let keyBytes = publicKeys[payload.keyId] else { throw EntitlementVerificationError.unknownKeyId }
        let key: Curve25519.Signing.PublicKey
        do { key = try Curve25519.Signing.PublicKey(rawRepresentation: keyBytes) }
        catch { throw EntitlementVerificationError.unknownKeyId }
        guard key.isValidSignature(signature, for: payloadBytes) else { throw EntitlementVerificationError.invalidSignature }
        guard payload.installationId == installationId else { throw EntitlementVerificationError.installationMismatch }
        if let expectedClubId, payload.clubId != expectedClubId { throw EntitlementVerificationError.unexpectedClub }
        guard payload.permittedPackIds.contains(expectedPackId) else { throw EntitlementVerificationError.unpermittedPack }
        guard payload.refreshAfter >= payload.issuedAt,
              payload.issuedAt <= now.addingTimeInterval(5 * 60),
              payload.entitlementType == (payload.perpetualAccess ? .perpetual : .term),
              (payload.perpetualAccess ? payload.expiresAt == nil && payload.graceUntil == nil : payload.expiresAt != nil && payload.graceUntil != nil)
        else { throw EntitlementVerificationError.inconsistentDates }
        if let expires = payload.expiresAt, let grace = payload.graceUntil, grace < expires { throw EntitlementVerificationError.inconsistentDates }
        if !allowExpiredReferenceAccess, let boundary = payload.graceUntil ?? payload.expiresAt, now > boundary {
            throw EntitlementVerificationError.expiredBeyondGrace
        }
        return VerifiedEntitlement(envelope: envelope, payload: payload)
    }
}

extension Data {
    init?(base64URLEncoded value: String) {
        guard !value.isEmpty, value.range(of: "[^A-Za-z0-9_-]", options: .regularExpression) == nil else { return nil }
        var base64 = value.replacingOccurrences(of: "-", with: "+").replacingOccurrences(of: "_", with: "/")
        base64 += String(repeating: "=", count: (4 - base64.count % 4) % 4)
        self.init(base64Encoded: base64)
    }
}

extension JSONDecoder {
    static var licensing: JSONDecoder {
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        return decoder
    }
}

extension JSONEncoder {
    static var licensing: JSONEncoder {
        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .iso8601
        return encoder
    }
}

File: android/app/src/main/java/com/lukematthews/syccourses/EntitlementVerifier.kt
package com.lukematthews.syccourses

import com.google.crypto.tink.subtle.Ed25519Verify
import kotlinx.serialization.json.Json
import java.time.Instant
import java.util.Base64

enum class EntitlementVerificationError { MALFORMED, UNKNOWN_SCHEMA, UNKNOWN_KEY, INVALID_SIGNATURE, INSTALLATION_MISMATCH, UNEXPECTED_CLUB, UNPERMITTED_PACK, INCONSISTENT_DATES, EXPIRED }
class EntitlementVerificationException(val reason: EntitlementVerificationError) : Exception(reason.name)

class EntitlementVerifier(
    private val publicKeys: Map<String, ByteArray>,
    private val json: Json = Json { ignoreUnknownKeys = true },
) {
    fun verify(
        envelope: EntitlementEnvelope,
        installationId: String,
        expectedClubId: String?,
        expectedPackId: String,
        now: Instant,
        allowExpiredReferenceAccess: Boolean = true,
    ): VerifiedEntitlement {
        val payloadBytes = decodeBase64Url(envelope.payload)
        val signature = decodeBase64Url(envelope.signature)
        val payload = runCatching { json.decodeFromString<EntitlementPayload>(payloadBytes.decodeToString()) }
            .getOrElse { throw EntitlementVerificationException(EntitlementVerificationError.MALFORMED) }
        if (payload.schemaVersion != 1) throw EntitlementVerificationException(EntitlementVerificationError.UNKNOWN_SCHEMA)
        val key = publicKeys[payload.keyId] ?: throw EntitlementVerificationException(EntitlementVerificationError.UNKNOWN_KEY)
        try { Ed25519Verify(key).verify(signature, payloadBytes) }
        catch (_: Exception) { throw EntitlementVerificationException(EntitlementVerificationError.INVALID_SIGNATURE) }
        if (payload.installationId != installationId) throw EntitlementVerificationException(EntitlementVerificationError.INSTALLATION_MISMATCH)
        if (expectedClubId != null && payload.clubId != expectedClubId) throw EntitlementVerificationException(EntitlementVerificationError.UNEXPECTED_CLUB)
        if (expectedPackId !in payload.permittedPackIds) throw EntitlementVerificationException(EntitlementVerificationError.UNPERMITTED_PACK)
        val issuedAt = parseInstant(payload.issuedAt)
        val refreshAfter = parseInstant(payload.refreshAfter)
        val expiresAt = payload.expiresAt?.let(::parseInstant)
        val graceUntil = payload.graceUntil?.let(::parseInstant)
        val correctType = payload.entitlementType == if (payload.perpetualAccess) "perpetual" else "term"
        val correctBoundaries = if (payload.perpetualAccess) expiresAt == null && graceUntil == null else expiresAt != null && graceUntil != null
        if (refreshAfter < issuedAt || issuedAt > now.plusSeconds(300) || !correctType || !correctBoundaries ||
            (expiresAt != null && graceUntil != null && graceUntil < expiresAt)) {
            throw EntitlementVerificationException(EntitlementVerificationError.INCONSISTENT_DATES)
        }
        if (!allowExpiredReferenceAccess && now > (graceUntil ?: expiresAt ?: Instant.MAX)) {
            throw EntitlementVerificationException(EntitlementVerificationError.EXPIRED)
        }
        return VerifiedEntitlement(envelope, payload, issuedAt, refreshAfter, expiresAt, graceUntil)
    }

    private fun decodeBase64Url(value: String): ByteArray {
        if (value.isEmpty() || !value.matches(Regex("[A-Za-z0-9_-]+"))) throw EntitlementVerificationException(EntitlementVerificationError.MALFORMED)
        return try { Base64.getUrlDecoder().decode(value) }
        catch (_: IllegalArgumentException) { throw EntitlementVerificationException(EntitlementVerificationError.MALFORMED) }
    }

    private fun parseInstant(value: String): Instant = runCatching { Instant.parse(value) }
        .getOrElse { throw EntitlementVerificationException(EntitlementVerificationError.INCONSISTENT_DATES) }
}
